keps/sig-auth/4193-bound-service-account-token-improvements/README.md

```diff
@@ -0,0 +1,622 @@
+passed along to the external service, thus removing the ability for a malicious actor to 'replay' a projected token
+from another Node.
+
+This will be implemented as an additional `node` entry in the private claims embedded into each JWT returned by the
+TokenRequest API, in a similar manner to how the ServiceAccount, Pod or Secret is referenced.
+
+Additionally, to provide a robust means of tracking token usage within the audit log we can embed a unique identifier for
+each token which is can then also be recorded in future audit entries made by this token.
+
+## Motivation
+
+### Goals
+
+* Embedding information about the Node that a pod is running on into signed JWTs.
+* Make it easier to track the actions a single token has taken, and cross-reference that back to the origin of the token
+  (via audit log inspection).
+* Provide a means of checking whether a Pod's token is associated with the same Node as it was associated with when the
+  initial TokenRequest was made (via an extra field that can be observed from the TokenReview API).
+
+### Non-Goals
+
+* Embedding requester information. This is discussed further in the alternatives
+  considered section, and a future KEP may revisit this.
+* Embedding information beyond the immutable Node name and UID into the token. We aim to mimic what is done with the ref fields
+  for secret, pod and serviceaccount (not introduce any additional properties).
+* Changing default behaviour of the SA authenticator to enforce the referenced Node object still exists.
+
+## Proposal
+
+### Embedding Pod's bound Node information in tokens
+
+The kube-apiserver will be extended to automatically embed the `name` and `uid` of the *Node* a Pod is associated
+with (via `spec.nodeName`) in generated tokens when a TokenRequest `create` call is serviced.
+
+As the 'pod' is already available in this area of code, which contains the `nodeName`, we will just need to plumb
+through a Getter for Node objects into the TokenRequest storage layer so the node's UID can be fetched, similar to
+what is done for pod & secret objects.
+
+### Allowing ServiceAccount tokens to be bound to a Node object
+
+Similar to how a token can be bound to a Pod or Secret object, we will also extend the TokenRequest API to allow
+binding directly to Node objects (without needing to bind to a Pod as well).
+
+This allows users to obtain a token that is tied specifically to the *Node* objects lifecycle, i.e. when the Node
+object is deleted, the token will be invalidated.
+
+### Extending TokenReview to verify tokens bound to Node objects
+
+The SA authenticator will be extended to check whether a token that is bound to a Node object is still valid, by
+first checking whether the Node object with the name given in the JWT still exists, and if it does, validating whether
+the UID of that Node is equal to the UID embedded in the token.
+
+Tokens bound to Pod objects will continue to only validate the referenced pod.
```

Comment:
Was re-reading this today and I wanted to double-check, I think this means:
- We won't check that the node exists for Pod-bound tokens, but
- Those tokens will still contain a validated node binding claim, which Kubelet will request when it requests the token.

Am I reading that correctly? @munnerz @liggitt

Reply:
The request from the node won't change, it will only bind to the pod as it does today. The apiserver already verifies the pod belongs to the requesting node in NodeRestriction admission. That node info looked up by the apiserver would now be added to the issued pod-bound tokens as node name/uid claims as well.

Reply:
When those pod bound tokens are presented and used to authenticate, only the pod referenced in the pod claim will be validated to still exist, as today